### DISCUSSION TITLE
Welcome to SmartEVSE-2 Discussions!

### DISCUSSION THREAD

**Discussion by mstegen** [SmartEVSE/SmartEVSE-2#19] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by mintgroen** [comment]
Hi Michael,

First of all, the SmartEVSE setup is happily charging my Hyundai Kona!

Do you have a feature list of SmartEVSE/Sensorboxfeatures that you are planning to develop in the near future?

I would really like to monitor the process and use more parameters that are available in my Domotica system to control the moment of charging and the charge current. The RS485 modbus is an option, but since there is an ESP32 idling in the sensorbox, that could be a more flexible option.

Looking forward to your reply!

Regards,
Maarten

**Reply by mstegen** [reply]
Hi Maarten, 

Plan is to use the ESP32 as a modbus master, so it can control all the connected SmartESVE controllers.
There will be a webpage hosted on the ESP, that you can use to configure all settings, and monitor / control the state of all controllers.
But unfortunately there's still a lot to do. 

Regards, 
Michael

**Reply by mintgroen** [reply]
Is there any news to mention?

A smart integration of the SmartEVSE into our smart homes would be very nice.

For example, I can't really use solar charging because then the SmartEVSE won't allow preconditioningy my plugged in EV on mains, at 6h30, obviously, the sun doesn't shine.

**Reply by IgorYbema** [reply]
And maybe share the code you are working on so the community can contribute on making this available?

**Comment by prjh** [comment]
Hello Michael,
I also want to have an external Modbus-client (Master) managing all my devices (PV-Inverter/battery, heat pump and smart meter and SmartEVSE!).
Is there a configuration, where I can access the SmartEVSE-2 as single Modbus server (slave). My first choice was to configure it as a slave (Node-1), but the SmartEVSE shows an error: "ERROR NO SERIAL COM - CHECK WIRING", 
... because it awaits a periodic master request.
Is there chance to simulate an extern SmartEVSE master? What ist the message, I have to create, whats the period? 

Kind regards 
Jörg

**Comment by prjh** [comment]
New status: I have updated to new revision 2.30!
I have studied the source code. For to control in the 1st step one EVSE, I will give a first trial with following setting:
Mode: Smart
LoadBalancing: Master (as Node-1)
For to modify the charging current, I will use the server address=1 (Master Node-1), with 'fc=0x10 (write-multiple-register) to the broadcast register 0x0020 with the actual nominal current value. In the current implementation this approach is restricted to non-master nodes. That this could work `Evse.c: line 2800` 
```
                        case 0x10: // (Write multiple register))
                            // 0x0020: Balance currents
                            if (Modbus.Register == 0x0020 && LoadBl > 1) {      // Message for Node(s)
                                Balanced[0] = (Modbus.Data[(LoadBl - 1) * 2] <<8) | Modbus.Data[(LoadBl - 1) * 2 + 1];
```
should be modified to:
```
                        case 0x10: // (Write multiple register))
                            // 0x0020: Balance currents              v--- ... also for master!
                            if (Modbus.Register == 0x0020 && LoadBl >= 1) {     // Message for Node(s)
                                Balanced[0] = (Modbus.Data[(LoadBl - 1) * 2] <<8) | Modbus.Data[(LoadBl - 1) * 2 + 1];

``` 
Great thanks, for that cool device and open source firmware!

**Comment by prjh** [comment]
Update:
To setup an external modbus client (master) for an external Modbus library for broadcast is mostly not supported (e.g. github.com/eModbus) and will produce timeouts due to the missing modbus response. 
So I decided to use a FC=x06 (write hold register) on staus register 0x0020 and set the Isum variable within the SOLAR mode an balancing as master. But the likely disabled mainmeter in the initial button based setting isn't possible. My EV-Meter was also set to address=10. But a second look shows that a disabled mainmeter is the same as using sensorbox as meter on addr=10. So the result is/was, the ISum value will be reseted periodically. Side effect is, that the second modbus master does inhibit broadcasts for some seconds. 
So, my initial idea to inject the mainmeter current meter value from an additional master (TCP/RTU gateway) isn't that easy an needs more changes to the code and FW architecture. 
At the current stage, I think a better way would be, to simulate an RTU based mainmeter for current readings of a SmartEVSE supported type on the RTU based EVSE local bus, which also has a buildin Modbus TCP-server and acts as a proxy to translate TCP side managment settings to  the local SmartEVSE subsystem without any implementation modifications(!).
Currently I use a Elfin WiFi TCP/RTU Modbus gateway from china. It is a nice and cheap device , but I do not have FW access - means I have to do it by myself.  
Congratulations - I have become some insides: SmartEVSE is a well thought out implementation for a smart, small EVSE system.

**Reply by bobosch** [reply]
Hi! Have you tried to write register 0x0002 ?
Register 0x0020 is a proprietary way between SmartEVSE ...

**Reply by prjh** [reply]
Hi @bobosch,
THX for your reply.
I just repeated your recommendation: writing "99" to 0x0002 for 9.9A does work, But reading 0x0002 again returns a 0.0A. But I do not have a connected EV (State-A). Could it be that this works only with a connected EV?

**Reply by bobosch** [reply]
Yes

**Comment by prjh** [comment]
Update:
I now have a better simulation environment. I have created two Modbus-Server applications which does simulate an SDM630 energy meter (EM). One for the mains and one for the EV side. 
![image](https://user-images.githubusercontent.com/15111616/148421495-43d7e214-cb5c-43ca-a873-9d439aceb76b.png)
The 'M5Atom Lite' ESP32 with RS485 interface does come with one button. I can now increase/decrease the available solar PV current value. This enables me to understand, check and validate the behaviour of the SmartEVSE. This is a prerequisite for my project for to externally control my EVSE infrastructure, where I want do a prioritization of different energy consumers, like local battery storage, heat pump, and EVSE coupled EV's. As the that nice device does have also have WiFi, I could communicate with it with an additional Modbus-TCP Server receiving the necessary energy meter system values - wihout a RS485 cable from my electrical disribution to my garage.
THX for that fine project. The more I get to know it now, the more I see the advantages of this smart solution.
